sklearn/utils/sparsefuncs.py

```diff
@@ -467,7 +467,8 @@ def count_nonzero(X, axis=None, sample_weight=None):
     elif axis == 1:
         out = np.diff(X.indptr)
         if sample_weight is None:
-            return out
+            # astype here is for consistency with axis=0 dtype
+            return out.astype('int')
```

Comment:
What's the purpose of it? Seems that it won't work:
```
import scipy.sparse as sp
from sklearn.utils.sparsefuncs import count_nonzero
X = sp.csr_matrix([[1, 0], [0, 1]])
print(count_nonzero(X, axis=0).dtype)
print(count_nonzero(X, axis=1).dtype)
print(((count_nonzero(X, axis=1)).astype('int')).dtype)
# int64
# int32
# int32
```

Reply:
Hmm.... Well it certainly fixed the doctest, which is what showed me the inconsistent behaviour. I admit it's a bit messy to include that fix here.

Are you on a 32-bit platform? I suppose I'm wrong to cast as 'int' here, and should be casting as whatever bincount returns (where is that specified?).

Reply:
Bincount uses numpy.intp. Is that not the same as int64 on your platform? If not, why then is `count_nonzero(X, axis=0)` int64?

Reply:
Note that this only happens on my Windows (not on my Linux).

> Bincount uses numpy.intp.

If it's the case (curious where do you find it), then the problem is that sometimes ``int`` is not the same as ``intp``? Maybe we can use ``intp``? I guess I'm on a 64-bit platform.
```
import numpy as np
print(np.dtype("int"))
print(np.dtype("intp"))
# int32
# int64
```
output of sklearn.show_versions()
```
System:
    python: 3.5.4 |Continuum Analytics, Inc.| (default, Aug 14 2017, 13:41:13) [MSC v.1900 64 bit (AMD64)]
   machine: Windows-10-10.0.16299-SP0
executable: C:\Anaconda2\envs\scikit_learn\python.exe

BLAS:
    macros: SCIPY_MKL_H=None, HAVE_CBLAS=None
cblas_libs: mkl_rt
  lib_dirs: C:\Anaconda2\Library\lib

Python deps:
       pip: 18.0
    pandas: 0.23.3
   sklearn: 0.21.dev0
     scipy: 1.1.0
setuptools: 40.0.0
    Cython: 0.28.5
     numpy: 1.15.1
```

Reply:
The problem is that it only happens on certain platforms. If you're sure that bincount use intp (apologies I'm not able to find related doc), maybe we can simply change int to intp here.